feat: add docs/concepts/storage with pricing info (#409)
near/docs: docs/concepts/storage.md

Take care, then, when designing your smart contracts to ensure that such attacks cost potential attackers more than it would be worth.

## btw, you can remove data to unstake some tokens

People familiar with the "immutable data" narrative about blockchains find this surprising. While it's true that an _indexing node_ will keep all data forever, _validating nodes_ (that is, the nodes run by most validators in the network) do not. Smart contracts can provide ways to delete data, and this data will be purged from most nodes in the network within a few [epochs](https://docs.near.org/docs/concepts/epoch).

Note that a call to your smart contract to remove data has an associated gas fee. Also, since deletions takes time, there is an upper limit on the amount of data you can remove in a single transaction, as your transaction must be able to fit in the one-second block time.

## How much does it cost?

> State staking is priced in an amount set by the network, which was initialized to **[9.09E19 yoctoNEAR per byte](https://github.com/nearprotocol/nearcore/blob/master/neard/res/genesis_config.json#L38)** and could be lowered (but not increased) in the future. NEAR's JSON RPC API provides [a way to query this initial setting](https://docs.near.org/docs/api/rpc-experimental#genesis-config), but does not yet provide a way to query the "live" configuration value.

Author: @bowenwang1996 is it true that `storage_amount_per_byte` will never be increased?

Reviewer: No

Author: It could be increased? What would happen to all the contracts with insufficient free balance to stake for the higher cost?

Reviewer: It depends on how we upgrade. For example we can make it such that the contract will be removed after some amount of time if no one funds it. Alternatively we can say that we will only allow transfer to this account until if has sufficient tokens to pay for storage.

Author: Still seems like something that could cause problems, even if mostly social problems. But thanks, I'll update the language.

Note in this example that it is the token holders and not your smart contract account which, [by default](https://docs.near.org/docs/concepts/gas#what-about-prepaid-gas), pay the gas fee.

## The "million worthless escrows" attack

Note that this can create an attack surface. To continue the example above, if naming an escrow costs a token holder close to nothing while costing the contract owner significantly more, then a malicious user can exploit the imbalance to make maintaining the contract prohibitively expensive.

Take care, then, when designing your smart contracts to ensure that such attacks cost potential attackers more than it would be worth.

## btw, you can remove data to unstake some tokens

People familiar with the "immutable data" narrative about blockchains find this surprising. While it's true that an _indexing node_ will keep all data forever, _validating nodes_ (that is, the nodes run by most validators in the network) do not. Smart contracts can provide ways to delete data, and this data will be purged from most nodes in the network within a few [epochs](https://docs.near.org/docs/concepts/epoch).

> Note that a call to your smart contract to remove data has an associated gas fee. Also, since deletions takes time, there is an upper limit on the amount of data you can remove in a single transaction, as your transaction must be able to fit in the one-second block time.

Reviewer 2: …in a one-second block time and due to computational limits per block. (don't know the source)

Reviewer: block time is not really one second. It is better to talk about gas limit here.

Author: @bowenwang1996 is the gas limit and the reason for it documented somewhere such as Nomicon?

---
id: storage
title: Storage Staking
sidebar_label: Storage Staking
---

> If you’re familiar with Ethereum’s pricing model, you may know that the protocol charges a fee (called "gas") once for a transaction, and this transaction takes into account the amount of data stored via that transaction. This essentially means that anyone can pay once to store permanent data on-chain. This is a poor economic design for at least two reasons: 1. The people running the network (miners, in the case of Ethereum 1) are not appropriately incentivized to store large amounts of data, since a gas fee far charged in the distant past can increase storage costs forever, and 2. The users of a smart contract are charged for the data they send to store in it, rather than charging the owner of the smart contract.

Reviewer 3: What's the target audience? Is it possible to explain to a dev who is not familiar with Ethereum? I think it's better to start with explaining what NEAR does and then say how it's different from Ehtereum.

> NEAR addresses this economic misalignment using a mechanism called storage staking (you may have heard that NEAR charges rent for storage – this has not been true [since March 2020](https://github.com/nearprotocol/nearcore/pull/2272)). In storage staking (sometimes called _state_ staking), the account that owns a smart contract must stake (or lock) tokens according to the amount of data stored in that smart contract, effectively reducing the balance of the contract's account.

Reviewer 3: If we start with this paragraph, this doc will be much more friendly for a non-blockchain but technically solid reader.

Author: Great idea, thanks! Rearranged.

## How does this align incentives?

> Storage-staked tokens are unavailable for other uses such as validation staking. This increases the yield that validators will receive. Learn more in [the economics whitepaper](https://near.org/papers/economics-in-sharded-blockchain/).

Reviewer 3: grammar nit: comma missing before "such as validation staking"

## When do tokens get staked?

On each incoming transaction that adds data.

Let's walk through an example:

1. You launch a [a custom token](https://examples.near.org/?q=fungible), deploying your token's smart contract to the account `mycoiniscoolerthanyours.near`
2. Token holders can name an escrow – an account which can transfer tokens on their behalf. Naming an escrow will require a call to your smart contract, and your smart contract will need to store this mapping of token-owner-to-escrow-account.

> 3. When this call comes in, NEAR will check that `mycoiniscoolerthanyours.near` has a large enough balance that it can stake an amount corresponding to this data addition. If it does not, the transaction will fail. Otherwise, these new tokens get locked for storage staking and the transaction completes.

Reviewer 3: can we use a different contract as an example? It can be confusing which tokens you are staking (NEAR vs mycoiniscoolerthanyours) Even switching to NFT will make it easier to grok.

Author: This cleaned it up a lot; great suggestion

Note in this example that it is the token holders and not your smart contract account which, [by default](https://docs.near.org/docs/concepts/gas#what-about-prepaid-gas), pay the gas fee.

## The "million worthless escrows" attack

Note that this can create an attack surface. To continue the example above, if naming an escrow costs a token holder close to nothing while costing the contract owner significantly more, then a malicious user can exploit the imbalance to make maintaining the contract prohibitively expensive.

Take care, then, when designing your smart contracts to ensure that such attacks cost potential attackers more than it would be worth.

## btw, you can remove data to unstake some tokens

> People familiar with the "immutable data" narrative about blockchains find this surprising. While it's true that an _indexing node_ will keep all data forever, _validating nodes_ (that is, the nodes run by most validators in the network) do not. Smart contracts can provide ways to delete data, and this data will be purged from most nodes in the network within a few [epochs](https://docs.near.org/docs/concepts/epoch).

Reviewer 3: possibly future improvement considerations: what is indexing node vs validating node?

Author: I could not find documentation for these concepts; I only know that these are possibilities because @bowenwang1996 said so in a call. Bowen, do you know if indexing nodes are documented anywhere?

Note that a call to your smart contract to remove data has an associated gas fee. Also, since deletions takes time, there is an upper limit on the amount of data you can remove in a single transaction, as your transaction must be able to fit in the one-second block time.

## How much does it cost?

> Storage staking is priced in an amount set by the network, which was initialized to **[9.09E19 yoctoNEAR per byte](https://github.com/nearprotocol/nearcore/blob/master/neard/res/genesis_config.json#L38)** and could be lowered (but not increased) in the future. NEAR's JSON RPC API provides [a way to query this initial setting](https://docs.near.org/docs/api/rpc-experimental#genesis-config), but does not yet provide a way to query the "live" configuration value.

Reviewer 4: Might be good to translate the cost to bytes per NEAR.

Someday this value will probably change, and when it does this document will be updated to include information about how to query the live version.

## Example cost breakdown

Let's walk through an example.

> A non-fungible token is one where each token is unique, which means each token has an ID. Amongst other things, such a contract must store a mapping from token IDs to owners' account addresses.

Reviewer 4: ```suggestion
A non-fungible token is unique, which means each token has its own ID. The contract must store a mapping from token IDs to owners' account ID.
```

> If such an NFT is used to track **1 million** tokens, how much storage will be required, just for that token-ID-to-owner mapping? And how many tokens will need to be staked for that storage?

Reviewer 4: ```suggestion
If such an NFT is used to track **1 million** tokens, how much storage will be required for the token-ID-to-owner mapping? And how many tokens will need to be staked for that storage?
```

* key: `t2o::0`
* value: `mytokeniscoolerthanyours.near`

So for 1 million tokens, here are all the things we need to add up and multiply by 1 million:

1. The prefix, `t2o`, will be serialized as three bytes in UTF-8, and the two colons will add another two. That's 5 bytes.
2. For an implementation where `TokenId` auto-increments, the values will be between `0` and `999999`, which makes the average length 5 bytes.
3. Let's assume well-formed NEAR `AccountId`s, and let's guess that NEAR Account IDs follow the approximate pattern of domain names, which [average about 10 characters](https://www.domainregistration.com.au/news/2013/1301-domain-length.php), plus a top-level name like `.near`. So a reasonable average to expect might be about 15 characters; let's keep our estimate pessimistic and say 25. This will equal 25 bytes, since NEAR account IDs must use characters from the ASCII set.

So:

    1_000_000 * (5 + 5 + 25)

35 million bytes. Multiplying by 9.09E19 yoctoNEAR per byte, we find that the `tokenToOwner` mapping will require staking 3.1815E27 yoctoNEAR, or Ⓝ3,181.5

> Note that you can get this down to Ⓝ2,999.7 just by changing the prefix from `t2o` to a single character

Reviewer 4: You can get rid of the whole prefix too, so -3 more bytes.

Author: Yeah, I may as well mention that!

> Note that you can get this down to Ⓝ2,999.7 just by changing the prefix from `t2o` to a single character

Reviewer 3: +100 to providing suggestions on how to minimize costs! Are there any other tips we can come up with at this point?

Author: There's a whole section on this at the bottom 😄

Author: ...which I now realize I failed to push before you reviewed! Sorry; I had it written but failed to push 😞
